pkg/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -852,6 +1001,108 @@ func (n *NetworkPolicyController) deleteService(old interface{}) {
+	klog.V(2).Infof("Processing Endpoints %s/%s UPDATE event", curEndpoints.Namespace, curEndpoints.Name)
+	// Find all ACNPs that are impacted by this Endpoints and reprocess them.
+	// Because Service and its Endpoints use the same Namespace and Name, pass
+	// Namespace and Name of Endpoints into filterACNPsByService
+	affectedACNPsByService := n.filterACNPsByService(curEndpoints.Namespace, curEndpoints.Name)
+	for _, acnp := range affectedACNPsByService {
+		n.reprocessACNP(acnp)
+	}
+}
+
+// deleteEndpoints receives Endpoints DELETE events and reprocesses the
+// ClusterNetworkPolicy with `toService` referred to the Service which uses this
+// Endpoints as backend.
+func (n *NetworkPolicyController) deleteEndpoints(oldObj interface{}) {
+	endpoints, ok := oldObj.(*v1.Endpoints)
+	if !ok {
+		tombstone, ok := oldObj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Error decoding object when deleting Endpoints, invalid type: %v", oldObj)
+			return
+		}
+		endpoints, ok = tombstone.Obj.(*v1.Endpoints)
+		if !ok {
+			klog.Errorf("Error decoding object tombstone when deleting Endpoints, invalid type: %v", tombstone.Obj)
+			return
+		}
+	}
+	defer n.heartbeat("deleteEndpoints")
+
+	// Find all ACNPs that are impacted by this Endpoints and reprocess them.
+	// Because Service and its Endpoints use the same Namespace and Name, pass
+	// Namespace and Name of Endpoints into filterACNPsByService
+	affectedACNPsByService := n.filterACNPsByService(endpoints.Namespace, endpoints.Name)
+	for _, acnp := range affectedACNPsByService {
+		n.reprocessACNP(acnp)
+	}
+}
+
+// filterACNPsByService gets all ClusterNetworkPolicy that will need to be
+// re-processed based on the "[Namespace]/[Name]" of an updated/deleted Service/Endpoints.
+func (n *NetworkPolicyController) filterACNPsByService(serviceNamespace string, serviceName string) []*secv1alpha1.ClusterNetworkPolicy {
+	var affectedPolicies []*secv1alpha1.ClusterNetworkPolicy
+	acnps, err := n.acnpInformer.Informer().GetIndexer().ByIndex(ServiceIndex, fmt.Sprintf("%s/%s", serviceNamespace, serviceName))
+	if err != nil {
+		klog.Errorf("Error fetching ClusterNetworkPolicies that have ToService rules: %v", err)
+		return nil
+	}
+	for _, acnpObj := range acnps {
+		acnp := acnpObj.(*secv1alpha1.ClusterNetworkPolicy)
+		affectedPolicies = append(affectedPolicies, acnp)
+	}
+	return affectedPolicies
+}
+
+// filterACNPsByServiceType gets all ClusterNetworkPolicy that will need to be
+// re-processed based on the serviceType when added/updated/deleted a Node.
+func (n *NetworkPolicyController) filterACNPsByServiceType(serviceType v1.ServiceType) []*secv1alpha1.ClusterNetworkPolicy {
+	var affectedPolicies []*secv1alpha1.ClusterNetworkPolicy
+	acnps, err := n.acnpInformer.Informer().GetIndexer().ByIndex(ServiceTypeIndex, string(serviceType))
+	if err != nil {
+		klog.Errorf("Error fetching ClusterNetworkPolicies with ServiceType:%s of Service in ToService rules: %v", serviceType, err)
```

Comment:
use structured looging for all new log messages

Reply:
Done.